### DISCUSSION TITLE
How can it perform a rolling update without downtime?

### DISCUSSION THREAD

**Discussion by rikuson** [socketio/socket.io#4857] [Q&A] [answered]
I'm hosting socket.io with a Redis stream adapter on ECS, and I'm encountering two issues when trying to perform a rolling update.  
I don't use sticky session and set `transports: ["websocket"]` .

> Connection state recovery is a feature which allows restoring a client's state after a temporary disconnection, including any missed packets.

Does this feature only apply to client-side disconnections? What if the server crashes and restarts? Is there a way to restore a client's state in this case?

Based on my investigation, it seems that the system treats reconnection as a new connection with a new socket ID and doesn't map the state to the reconnected client.

Additionally, during the rolling out of pods, the server triggers events: "leave-room", "disconnect" and "delete-room" with showing the log below.

```
socket.io:client client close with reason transport close
```

There is my application code that erases some states from Redis in "disconnect" event handler.  
Ideally, pods die before triggering the "disconnect" event.

Is there a standard solution for these issues?

**Comment by darrachequesne** [comment, marked as answer]
Hi!

> Does this feature only apply to client-side disconnections?

No, it also works when the Socket.IO server is cleanly stopped:

```js
process.on('SIGTERM', () => {
  io.close();
});
```

The connections will be closed with a `server shutting down` reason, which is in the list of recoverable reasons:

https://github.com/socketio/socket.io/blob/9a2a83fdd42faa840d4f11fd223349e5d8e4d52c/lib/socket.ts#L50-L57

So the connection state will be persisted in the store (if the process has enough time to do so) and restored upon reconnection.

**Reply by rikuson** [reply]
Thank you!

> No, it also works when the Socket.IO server is cleanly stopped:

Right, calling close method kept the state.

> There is my application code that erases some states from Redis in "disconnect" event handler.

For this problem, maybe I can do something in the SIGTERM handler.
